### PR TITLE
Cascading Location Spinner Implementation

### PR DESCRIPTION
**Pull request by hamza-vd**
See Issue #739

### CODE REVIEW

File: opensrp-anc/src/main/java/org/smartregister/anc/library/constants/AncFormConstants.java [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+        public static final String FACILITY = "health_facility";
+        public static final String VILLAGE = "village";
+    }
+
+    public static final class FormKeyConstants {
+        public static final String SUB_TYPE = "sub_type";
+        public static final String LOCATION_SUB_TYPE = "location";
+    }
+
+}
```
Comment (dubdabasoduba) [inline opensrp-anc/src/main/java/org/smartregister/anc/library/constants/AncFormConstants.java:18]: I think we already have other constant files in the app. Please move these classes in there.

File: opensrp-anc/src/main/java/org/smartregister/anc/library/presenter/ANCJsonFormFragmentPresenter.java [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+                JSONObject field = JsonFormUtils.getFieldJSONObject(formFragment.getStep(STEP1).getJSONArray(FIELDS), key);
+                String parentLocationId = field.getString(VALUE);
+
+                if (key.equals(PROVINCE)) {
+                    populateLocationSpinner(parentLocationId, DISTRICT, descendants.get(DISTRICT));
+                } else if (key.equals(DISTRICT)) {
+                    populateLocationSpinner(parentLocationId, SUB_DISTRICT, descendants.get(SUB_DISTRICT));
+                } else if (key.equals(SUB_DISTRICT)) {
+                    populateLocationSpinner(parentLocationId, FACILITY, descendants.get(FACILITY));
+                } else if (key.equals(FACILITY) && position > -1) {
```
Comment (dubdabasoduba) [inline opensrp-anc/src/main/java/org/smartregister/anc/library/presenter/ANCJsonFormFragmentPresenter.java:73]: Please use https://commons.apache.org/proper/commons-lang/apidocs/org/apache/commons/lang3/StringUtils.html for this `equals` check

File: opensrp-anc/src/main/java/org/smartregister/anc/library/util/Utils.java [hunk 3]
```diff
@@ -974,4 +977,80 @@ private final void addEmptyLine(Document layoutDocument, int number) {
...
+            case "PROVINCE":
+                return province != null ? province.getId() : "";
+            case "DISTRICT":
+                return district != null ? district.getId() : "";
+            case "SUBDISTRICT":
+                return subDistrict != null ? subDistrict.getId() : "";
+            case "HEALTH_FACILITY":
+            case "FACILITY":
+                return facility != null ? facility.getId() : "";
+            case "VILLAGE":
```
Comment (dubdabasoduba) [inline opensrp-anc/src/main/java/org/smartregister/anc/library/util/Utils.java:1033]: Please move this to a constants file

File: opensrp-anc/src/main/java/org/smartregister/anc/library/util/Utils.java [hunk 4]
```diff
@@ -974,4 +977,80 @@ private final void addEmptyLine(Document layoutDocument, int number) {
...
+        String id = location.getProperties().getName().toLowerCase().trim()
+                .replace(" ", "_")
+                .replace("(", "")
+                .replace(")", "")
+                .replace("-", "_")
+                .replace(":", "_")
+                .replace("'", "")
+                .replace("’", "_");
+
+        int identifier = jsonFormView.getResources().getIdentifier(id, "string", jsonFormView.getApplicationContext().getPackageName());
```
Comment (dubdabasoduba) [inline opensrp-anc/src/main/java/org/smartregister/anc/library/util/Utils.java:1049]: Use a constant for the `string` string literal

File: opensrp-anc/src/main/java/org/smartregister/anc/library/util/Utils.java (hunk 3 above)
Comment (dubdabasoduba) [inline opensrp-anc/src/main/java/org/smartregister/anc/library/util/Utils.java:1033]: Please move this to a constants file

File: opensrp-anc/src/main/java/org/smartregister/anc/library/widget/ANCSpinnerFactory.java [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+            Timber.e(e);
+        }
+
+        return super.getViewsFromJson(stepName, context, formFragment, jsonObject, listener, popup);
+    }
+
+    protected void populateProvince(JSONObject jsonObject) {
+        List<LocationTag> tags = Utils.getLocationTagsByTagName("Country");
+        // TODO: Remove static country ID, current implementation is only done to select proper country in case of more than one countries (dummy data) in locations
+        String countryId = (tags != null && tags.size() > 0) ? ((tags.size() > 1) ? "02ebbc84-5e29-4cd5-9b79-c594058923e9" : tags.get(0).getLocationId()) : "";
```
Comment (dubdabasoduba) [inline opensrp-anc/src/main/java/org/smartregister/anc/library/widget/ANCSpinnerFactory.java:123]: Please remember to mos the static country ID to use configs from the `app.properties`. also, remember to check when it's set and log any errors appropriately

File: opensrp-anc/src/main/java/org/smartregister/anc/library/widget/ANCSpinnerFactory.java [hunk 6]
```diff
@@ -0,0 +1,192 @@
...
+            }
+        } catch (JSONException e) {
+            Timber.e(e);
+        }
+
+        return super.getViewsFromJson(stepName, context, formFragment, jsonObject, listener, popup);
+    }
+
+    protected void populateProvince(JSONObject jsonObject) {
+        List<LocationTag> tags = Utils.getLocationTagsByTagName("Country");
```
Comment (dubdabasoduba) [inline opensrp-anc/src/main/java/org/smartregister/anc/library/widget/ANCSpinnerFactory.java:121]: Move the string literals to the constants file

**Review by dubdabasoduba — CHANGES_REQUESTED** [review]

**Review by dubdabasoduba — APPROVED** [review]

**Review by dubdabasoduba — CHANGES_REQUESTED** [review]
Mistakenly approved Please address comments first

File: opensrp-anc/src/main/java/org/smartregister/anc/library/constants/AncFormConstants.java (hunk 1 above)
Comment (hamza-vd) [reply opensrp-anc/src/main/java/org/smartregister/anc/library/constants/AncFormConstants.java]: This is resolved

File: opensrp-anc/src/main/java/org/smartregister/anc/library/presenter/ANCJsonFormFragmentPresenter.java (hunk 2 above)
Comment (hamza-vd) [reply opensrp-anc/src/main/java/org/smartregister/anc/library/presenter/ANCJsonFormFragmentPresenter.java]: Done

File: opensrp-anc/src/main/java/org/smartregister/anc/library/util/Utils.java (hunk 4 above)
Comment (hamza-vd) [reply opensrp-anc/src/main/java/org/smartregister/anc/library/util/Utils.java]: Moved to constant

File: opensrp-anc/src/main/java/org/smartregister/anc/library/util/Utils.java (hunk 3 above)
Comment (hamza-vd) [reply opensrp-anc/src/main/java/org/smartregister/anc/library/util/Utils.java]: Done

File: opensrp-anc/src/main/java/org/smartregister/anc/library/widget/ANCSpinnerFactory.java (hunk 5 above)
Comment (hamza-vd) [reply opensrp-anc/src/main/java/org/smartregister/anc/library/widget/ANCSpinnerFactory.java]: Country ID is now referenced from `app.properties`

File: opensrp-anc/src/main/java/org/smartregister/anc/library/widget/ANCSpinnerFactory.java (hunk 6 above)
Comment (hamza-vd) [reply opensrp-anc/src/main/java/org/smartregister/anc/library/widget/ANCSpinnerFactory.java]: Done

**Review by dubdabasoduba — APPROVED** [review]

File: opensrp-anc/src/main/java/org/smartregister/anc/library/presenter/ANCJsonFormFragmentPresenter.java [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+        JSONObject field = new JSONObject();
+        JSONArray codes = new JSONArray();
+        JSONArray values = new JSONArray();
+
+        for (Location location : locations) {
+
+            codes.put(location.getId());
+            String locationName = Utils.getLocationLocalizedName(location, jsonFormView);
+
+            // values.put(location.getProperties().getName());
```
Comment (SebaMutuku) [inline opensrp-anc/src/main/java/org/smartregister/anc/library/presenter/ANCJsonFormFragmentPresenter.java:128]: Can you remove dead code?

File: opensrp-anc/src/main/java/org/smartregister/anc/library/util/Utils.java [hunk 8]
```diff
@@ -974,4 +977,81 @@ private final void addEmptyLine(Document layoutDocument, int number) {
...
+        } catch (JSONException e) {
+            Timber.e(e, "Error loading current location");
+        } catch (Exception e) {
+            Timber.e(e, e.getMessage());
+        }
+
+        return currentLocation;
+    }
+
+    private static String getDefaultLocation(String level, String villageId) {
```
Comment (SebaMutuku) [inline opensrp-anc/src/main/java/org/smartregister/anc/library/util/Utils.java:1016]: Please add javadocs for classes or methods you create

**Review by SebaMutuku — COMMENTED** [review]
Kindly see my comments for the above

File: opensrp-anc/src/main/java/org/smartregister/anc/library/util/Utils.java (hunk 8 above)
Comment (hamza-vd) [reply opensrp-anc/src/main/java/org/smartregister/anc/library/util/Utils.java:1024]: Added

File: opensrp-anc/src/main/java/org/smartregister/anc/library/presenter/ANCJsonFormFragmentPresenter.java (hunk 7 above)
Comment (hamza-vd) [reply opensrp-anc/src/main/java/org/smartregister/anc/library/presenter/ANCJsonFormFragmentPresenter.java]: Resolved

**Review by dubdabasoduba — APPROVED** [review]

**Pull request merged**